File: adapterpattern/adapterpattern.py
from abc import ABC, abstractmethod

# adaptee
class WeighingMachine(ABC):
    
    @abstractmethod
    def getWeightInPound(self):
        pass

class WeighingMachineForBabies(WeighingMachine):
    def __init__(self, weight):
        self.weight = weight

    def getWeightInPound(self):
        return self.weight
    
# adapter
class WeighingMachineAdapter(ABC):
    
    @abstractmethod
    def getWeightInKg(self):
        pass

class WeighingMachineAdapterImpl(WeighingMachineAdapter):

    def __init__(self, weighingMachine):
        self.weighingMachine = weighingMachine
    
    def getWeightInKg(self):
        weightInPound = self.weighingMachine.getWeightInPound()

        # Convert it to KGs
        weightInKg = weightInPound * 0.453592
        return weightInKg

# client
if __name__ == '__main__':
    weighingMachineAdapter = WeighingMachineAdapterImpl(WeighingMachineForBabies(50))
    print(weighingMachineAdapter.getWeightInKg())

File: adapterpattern/adapterpattern.js
// ===== Adaptee =====
class WeightMachine {
  getWeightInPound() {
    throw new Error('Abstract method "getWeightInPound()" must be implemented');
  }
}

class WeightMachineForBabies extends WeightMachine {
  getWeightInPound() {
    const weight = 56; // example weight
    return parseInt(weight, 10);
  }
}

// ===== Adapter Interface =====
class WeightMachineAdapter {
  getWeightInKg() {
    throw new Error('Abstract method "getWeightInKg()" must be implemented');
  }
}

// ===== Adapter Implementation =====
class WeightMachineAdapterImpl extends WeightMachineAdapter {
  constructor(weightMachine) {
    super();
    this.weightMachine = weightMachine;
  }

  getWeightInKg() {
    const weightInPound = this.weightMachine.getWeightInPound();
    const weightInKg = weightInPound * 0.45;
    return weightInKg;
  }
}

// ===== Client Code =====
const weightMachineAdapter = new WeightMachineAdapterImpl(
  new WeightMachineForBabies()
);
console.log(`Weight in Kg: ${weightMachineAdapter.getWeightInKg()}`);
